### PR TITLE
Add a comments section to the website

### PR DESCRIPTION
**Pull request by carolksun**
This PR adds a comments section to the portfolio website where visitors can 
input comments and they are displayed on the page. The visitor can also change 
the number of comments being shown and delete any comments. The comments are 
stored in Datastore for persistent data storage.

### CODE REVIEW

File: portfolio/pom.xml [hunk 1]
```diff
@@ -34,11 +46,10 @@
         <artifactId>appengine-maven-plugin</artifactId>
         <version>2.2.0</version>
         <configuration>
-          <!-- TODO: set project ID. -->
-          <deploy.projectId>YOUR_PROJECT_ID_HERE</deploy.projectId>
+          <deploy.projectId>cksun-step-2020</deploy.projectId>
           <deploy.version>1</deploy.version>
```
Comment (razvanm) [inline portfolio/pom.xml]: It might not be obvious but the version can be any string that is valid in DNS names. So you can use something like `pr-3` to indicates from what PR the version is coming from.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -13,20 +13,64 @@
...
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;  
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Isn't the `TODO` addressed in this PR?

File: portfolio/src/main/webapp/index.html [hunk 3]
```diff
@@ -3,34 +3,45 @@
...
+      <br>
+      <div id="fact-container" class="center"></div>
+
+      <h2>Comments Section</h2>
+      <form method="POST" action="/data">
+		<textarea rows="4" cols="50" name="text">Leave a comment!</textarea>
+        <br>
+        <button>Submit</button>
+      </form>
+      <p>
```
Comment (razvanm) [inline portfolio/src/main/webapp/index.html:43]: I don't see the `</p>` for this `<p>`.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -13,20 +13,64 @@
...
+            String text = (String) entity.getProperty("text");
+            long timestamp = (long) entity.getProperty("timestamp");
+
+            Comment comment = new Comment(id, text, timestamp);
+            comments.add(comment);
+        }
+
+        Gson gson = new Gson();
+
+        response.setContentType("application/json;");
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Is the `;` at the end necessary?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -13,20 +13,64 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The Google Java style guide doesn't like wildcard imports.

Reference: https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -13,20 +13,64 @@
...
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        List<Comment> comments = new ArrayList<>();
+        for (Entity entity : results.asIterable()) {
+            long id = entity.getKey().getId();
+            String text = (String) entity.getProperty("text");
+            long timestamp = (long) entity.getProperty("timestamp");
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: No space before entity. Also, would be good to indicate in a comment the data model the doGet and doPost use for Datastore.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -13,20 +13,64 @@
...
 public class DataServlet extends HttpServlet {
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
-  }
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I think `Comment`, `timestamp` and `text` should be constants to allow the compiler catch any typos.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import com.google.appengine.api.datastore.KeyFactory;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.*;  
+
+/** Servlet that deletes the entity that has the corresponding id.*/
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: I think there should be a space after `.`.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Forgot to delete it!

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (carolksun) [reply portfolio/src/main/webapp/index.html:43]: I was following the google style guide for HTML (https://google.github.io/styleguide/htmlcssguide.html#Optional_Tags)
where they say to omit optional tags. I looked it up and apparently, `</p>` is optional so I deleted those tags from the code

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I added them as String constants

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Deleted the space and added the comment

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Removed the * import and imported necessary packages

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Removed!

File: portfolio/pom.xml [hunk 9]
```diff
@@ -18,10 +18,22 @@
 
   <dependencies>
     <dependency>
-      <groupId>javax.servlet</groupId>
-      <artifactId>javax.servlet-api</artifactId>
-      <version>4.0.1</version>
-      <scope>provided</scope>
+        <groupId>javax.servlet</groupId>
```
Comment (yinggao) [inline portfolio/pom.xml]: Prefer to keep the same indent through the entire file. For example the previous content has 2 spaces indent while here has 4 spaces.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -13,20 +13,77 @@
...
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+    private final String com = "Comment";
+    private final String ts = "timestamp";
+    private final String txt = "text";
+
+    /** Store each text input from the server as a Comment class, which has 
+        properties id, text, and timestamp. */
```
Comment (yinggao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Multi-line comments usually starts with *.
e.g. https://google.github.io/styleguide/javaguide.html#s4.8.6-comments

File: portfolio/src/main/webapp/background.html [hunk 11]
```diff
@@ -3,34 +3,31 @@
...
+      </ul>
     </div>
-    <h1> Background </h1>
+    
+    <h1>Background</h1>
     <p> 
-      I am a computer science major at Caltech, expecting to graduate in June 2022. <br> 
-      So far, some of the theoretical CS classes I have taken are algorithms, decidability and tractability, and learning systems. <br> 
-      Some of the practical CS classes I've taken are data structures, software design, and machine learning. <br> 
+      I am a computer science major at Caltech, expecting to graduate in June 2022.<br>
```
Comment (yinggao) [inline portfolio/src/main/webapp/background.html]: This text is getting long, prefer to break it at 80 characters. 
https://developers.google.com/style/html-formatting#line-length

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -13,19 +13,70 @@
...
+        const messageListElement = document.getElementById('comments-container');
+        messageListElement.innerHTML = "";
+        messages.forEach((comment) => {
+            messageListElement.appendChild(createCommentElement(comment));
+        });
+    });
+}
+
+/**
+ * Create comment on the page with a delete button
```
Comment (yinggao) [inline portfolio/src/main/webapp/script.js]: Missing dot at the end?

File: portfolio/pom.xml (hunk 9 above)
Comment (carolksun) [reply portfolio/pom.xml]: Fixed the indentation!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Added * for second line

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: fixed!

File: portfolio/src/main/webapp/background.html (hunk 11 above)
Comment (carolksun) [reply portfolio/src/main/webapp/background.html]: fixed!

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (yinggao) [reply portfolio/src/main/webapp/index.html:43]: Another thing I noticed from the style guild is "For consistency and simplicity reasons it’s best served omitting all optional tags, not just a selection." In our case, for consistency it might be simpler to keep the optional tags in my opinion.

**Review by yinggao — COMMENTED** [review]
LGTM in general. I'd like to wait for Razvan's review and wanna see what's his take on omitting the optional tags.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+/** An item on a todo list. */
+public final class Comment {
+
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: I don't think this empty line is necessary.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+/** An item on a todo list. */
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: This class seem to describe a comment and not an item on a todo list.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 15]
```diff
@@ -13,20 +13,78 @@
...
+/** Servlet that handles comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+    private final String com = "Comment";
+    private final String ts = "timestamp";
+    private final String txt = "text";
+
+    /** Store each text input from the server as a Comment class, which has 
+      * properties id, text, and timestamp. 
+      */
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: According to [1] this should look like this:

```
    /**
     * Store each text input from the server as a Comment object, which has 
     * properties id, text, and timestamp. 
     */
```

The text of a comment is a bit confusing (who is the server? where the data is stored?). As far as I can tell the function is producing a JSON response containing all entities of a particular kind stored in Datastore [2].

[1] https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line
[2] https://cloud.google.com/datastore/docs/concepts/entities

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 16]
```diff
@@ -13,20 +13,78 @@
...
+import java.util.List;
+
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+    private final String com = "Comment";
+    private final String ts = "timestamp";
+    private final String txt = "text";
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Based on [1] the names more suitable names for the constants would be something like `COMMENT_KIND`, `TIMESTAMP_PROPERTY` and `TEXT_PROPERTY`.

[1] https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.*;  
+
+/** Servlet that deletes the entity that has the corresponding id. */
+@WebServlet("/delete-data")
+public class DeleteDataServlet extends HttpServlet {
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        long id = Long.parseLong(request.getParameter("id"));
+        Key commentEntityKey = KeyFactory.createKey("Comment", id);
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: We should use the constant for the `Comment` string literal.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.appengine.api.datastore.KeyFactory;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.*;  
+
+/** Servlet that deletes the entity that has the corresponding id. */
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: In normal text the word 'identifier' is typically written as 'ID' and not 'id'.

A relevant discussion: https://english.stackexchange.com/questions/101248/how-should-the-abbreviation-for-identifier-be-capitalized

Also: https://www.merriam-webster.com/dictionary/id

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 19]
```diff
@@ -13,20 +13,78 @@
...
+        long timestamp = System.currentTimeMillis();
+
+        Entity commentEntity = new Entity(com);
+        commentEntity.setProperty(txt, text);
+        commentEntity.setProperty(ts, timestamp);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(commentEntity);
+
+        response.sendRedirect("/index.html");
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: Is this necessary considering that we are using the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)?

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -13,19 +13,70 @@
...
+        messages.forEach((comment) => {
+            messageListElement.appendChild(createCommentElement(comment));
+        });
+    });
+}
+
+/**
+ * Create comment on the page with a delete button.
+ * After a comment is deleted, the comments are reloaded with the correct 
+ * number of comments shown.
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js]: Please reflow.

File: portfolio/src/main/webapp/script.js [hunk 21]
```diff
@@ -13,19 +13,70 @@
...
+
+    // Add it to the page.
+    const factContainer = document.getElementById('fact-container');
+    factContainer.innerText = fact;
+}
+
+/**
+ * Load a specified number of comments to the comments section.
+ * If the specified number is greater than the total number of comments, all
+ * comments are shown.
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js]: Please reflow.

File: portfolio/src/main/webapp/script.js [hunk 22]
```diff
@@ -13,19 +13,70 @@
...
-
-  // Add it to the page.
-  const factContainer = document.getElementById('fact-container');
-  factContainer.innerText = fact;
+    // List of facts to choose from.
+    const facts =
+        ["I've lived in LA my whole life", 
+        "I have surpassed level 3000 on Candy Crush Saga", 
+        "I have an older sister", 
+        "My favorite fruit is watermelon"];
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js]: A bit nicer form:

```
const facts = [
  "I've lived in LA my whole life", 
  "I have surpassed level 3000 on Candy Crush Saga", 
  "I have an older sister", 
  "My favorite fruit is watermelon",
];
```

File: portfolio/src/main/webapp/script.js [hunk 23]
```diff
@@ -13,19 +13,70 @@
 // limitations under the License.
 
 /**
- * Adds a random greeting to the page.
+ * Adds a random fact to the page.
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js]: Is 'Adds' accurate? As far as I can tell the fact is replaced.

File: portfolio/src/main/webapp/script.js [hunk 24]
```diff
@@ -13,19 +13,70 @@
...
+    });
+}
+
+/**
+ * Create comment on the page with a delete button.
+ * After a comment is deleted, the comments are reloaded with the correct 
+ * number of comments shown.
+ */
+function createCommentElement(comment) {
+    const coElement = document.createElement('li');
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js]: The variable name like 'coElement' is unusual for JavaScript. How about 'commentElement'?

File: portfolio/src/main/webapp/script.js [hunk 25]
```diff
@@ -13,19 +13,70 @@
...
+        });
+    });
+}
+
+/**
+ * Create comment on the page with a delete button.
+ * After a comment is deleted, the comments are reloaded with the correct 
+ * number of comments shown.
+ */
+function createCommentElement(comment) {
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js:61]: Create UI elements using `document.createElement` can be hard to follow. Might be good to give an example of what's the expected output.

**Review by razvanm — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 13 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Fixed!

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 14 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: I deleted the comment about the class since it seemed self explanatory that it was a comment with id, text, and timestamp

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 15 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: rewrote the comment

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 16 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed!

File: portfolio/src/main/webapp/script.js (hunk 25 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js:61]: added it to the comment

File: portfolio/src/main/webapp/script.js (hunk 24 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: fixed!

File: portfolio/src/main/webapp/script.js (hunk 23 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: I changed "adds" to "displays"

File: portfolio/src/main/webapp/script.js (hunk 22 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: fixed!

File: portfolio/src/main/webapp/script.js (hunk 21 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: done!

File: portfolio/src/main/webapp/script.js (hunk 20 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: done!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 18 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: fixed!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 17 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: added constants for "Comment" and "id"

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 19 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: From what I gathered from the tutorials, we need to put the comments into Datastore so the comments aren't deleted every time the server is reloaded

File: portfolio/src/main/webapp/script.js [hunk 26]
```diff
@@ -12,16 +12,15 @@
...
 
-/**
- * Adds a random fact to the page.
- */
+/** Displays a random fact to the page. */
 function addRandomFact() {
     // List of facts to choose from.
-    const facts =
-        ["I've lived in LA my whole life", 
+    const facts =[
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js]:
```suggestion
    const facts = [
```

File: portfolio/src/main/webapp/script.js [hunk 27]
```diff
@@ -50,13 +48,17 @@ function loadComments() {
...
- * Create comment on the page with a delete button.
- * After a comment is deleted, the comments are reloaded with the correct 
- * number of comments shown.
+ * Create comment on the page with a delete button. After a comment is deleted, 
+ * the comments are reloaded with the correct number of comments shown.
+ * This functions creates HTML element that has the format:
+ * <li class="comment">
+ *   <span>COMMENT_TEXT</span>
+ *   <button>Delete</button>
+ * </li>
```
Comment (razvanm) [inline portfolio/src/main/webapp/script.js]:
```suggestion
 * Create comment on the page with a delete button. After a comment is deleted, 
 * the comments are reloaded with the correct number of comments shown.
 *
 * This functions creates an HTML element that has the format:
 *
 *     <li class="comment">
 *       <span>COMMENT_TEXT</span>
 *       <button>Delete</button>
 *     </li>
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 19 above)
Comment (razvanm) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: I'm not following. I would expect the fetch to not cause the page to reload. Is that not the case?

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 28]
```diff
@@ -14,9 +14,7 @@
 
 package com.google.sps.data;
 
-/** An item on a todo list. */
 public final class Comment {
```
Comment (yinggao) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:21]: I'd still add some comments for the class. Although it may seem redundant, in reality, many code search or editor will have the function that shows the comments when you hover over the class name. That's when the comment is especially useful.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.*;  
+
+/** Servlet that deletes the entity that has the corresponding ID. */
+@WebServlet("/delete-data")
+public class DeleteDataServlet extends HttpServlet {
+    private final String COMMENT_KIND = "Comment";
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: We want to define `COMMENT_KIND` and the other constants only once. You might need to define a separate class to do that.

**Review by razvanm — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 26 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: fixed!

File: portfolio/src/main/webapp/script.js (hunk 27 above)
Comment (carolksun) [reply portfolio/src/main/webapp/script.js]: fixed the indentation and spacing

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 29 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: I added a StringConstants.java class and imported it into the DataServlet and the DeleteDataServlet

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 19 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:93]: I tried commenting out the redirect to index.html page and when I submit the comment, it redirects to a /data blank page which is why the redirect is necessary

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 28 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:21]: Added a comment!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 30]
```diff
@@ -14,6 +14,7 @@
 
 package com.google.sps.servlets;
 import com.google.sps.data.Comment;
+import static com.google.sps.data.StringConstants.*;
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Do we need the `*`? The Java style guide doesn't like that.

Reference: https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports

Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed!

File: portfolio/pom.xml [hunk 31]
```diff
@@ -34,11 +46,12 @@
         <artifactId>appengine-maven-plugin</artifactId>
         <version>2.2.0</version>
         <configuration>
-          <!-- TODO: set project ID. -->
-          <deploy.projectId>YOUR_PROJECT_ID_HERE</deploy.projectId>
-          <deploy.version>1</deploy.version>
+          <address>0.0.0.0</address>
```
Comment (razvanm) [inline portfolio/pom.xml]: The `address` doesn't actually do anything so I recommend remove it because it gives the impression that is doing something.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 32]
```diff
@@ -0,0 +1,28 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+/** A comment that stores the id, text, and timestamp for use in Datastore*/
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]:
```suggestion
/** A comment is a container that stores data retrieved from Datastore (ID, text, and timestamp) and returned to the user in JSON format. */
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 33]
```diff
@@ -13,20 +13,81 @@
...
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data */
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
/** Servlet that handles the comments data. */
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 34]
```diff
@@ -13,20 +13,81 @@
...
 public class DataServlet extends HttpServlet {
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
-  }
+    /** 
+     * Retrieves the (limit number) of most recent comments from Datastore
+     * and produces a JSON response containing these comments
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
     * and produces a JSON response containing these comments.
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 35]
```diff
@@ -13,20 +13,81 @@
...
+        }
+
+        Gson gson = new Gson();
+
+        response.setContentType("application/json");
+        response.getWriter().println(gson.toJson(comments));
+    }
+
+    /** 
+     * Stores the comment into datastore and redirect user back to home page
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
     * Stores the comment into Datastore and redirect user back to home page 
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 36]
```diff
@@ -13,20 +13,81 @@
...
+        Gson gson = new Gson();
+
+        response.setContentType("application/json");
+        response.getWriter().println(gson.toJson(comments));
+    }
+
+    /** 
+     * Stores the comment into datastore and redirect user back to home page 
+     * after submission. datastore.put() directs to /data page so the redirect
+     * prevents the user from reaching the /data page.
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This is not accurate: as we discussed, `datastore.put()` is not the call responsible for the redirect. The reason we need a redirect is because we use a `form` in the HTML file.

**Review by razvanm — CHANGES_REQUESTED** [review]

File: portfolio/pom.xml (hunk 31 above)
Comment (carolksun) [reply portfolio/pom.xml]: deleted! I added it when I was debugging why my maven appengine failed to build but turns out I just needed to restart the shell

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 36 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I changed the comment

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 37]
```diff
@@ -14,7 +14,10 @@
 
 package com.google.sps.data;
 
-/** A comment that stores the id, text, and timestamp for use in Datastore*/
+/** 
+ * A comment is a container that stores data retrieved from Datastore (ID, text,
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]:
```suggestion
 * A Comment is a container that stores data retrieved from Datastore (ID, text, 
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 38]
```diff
@@ -71,9 +71,9 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     }
 
     /** 
-     * Stores the comment into datastore and redirect user back to home page 
-     * after submission. datastore.put() directs to /data page so the redirect
-     * prevents the user from reaching the /data page.
+     * Stores the comment into Datastore and redirect user back to home page 
+     * after submission. The form to submit the comment on the website directs 
+     * to /data page so the redirect prevents the user from reaching the /data page.
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
     * to /data page so the redirect prevents the user from getting stuck on the empty /data page.
```

Also, please reflow the text.

**Review by razvanm — CHANGES_REQUESTED** [review]
Please bring the PR description in the final form (we don't want ephemeral URLs in it for example).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 39]
```diff
@@ -13,20 +13,81 @@
...
+            comments.add(comment);
+        }
+
+        Gson gson = new Gson();
+
+        response.setContentType("application/json");
+        response.getWriter().println(gson.toJson(comments));
+    }
+
+    /** Stores the comment into Datastore and redirect user back to home page
```
Comment (yinggao) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Prefer format:
```
/**
 * Stores ....
 * ...
 */
```
where the sentence starts on the second line.

Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed!

Comment (yinggao) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This seems didn't change.

Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I seem to have changed one of the comments but not all of them. The next commit fixes the comment formatting in all the java files

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 40]
```diff
@@ -29,7 +29,9 @@
 import java.util.ArrayList;
 import java.util.List; 
 
-/** Servlet that deletes the entity that has the corresponding ID. */
```
Comment (yinggao) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: (No change required) You don't have to change the comment format here, one line also works. Just when it's a multi-line comment, the sentence should start at the second line.

**Review by yinggao — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java (hunk 40 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: oh okay, I changed it to be consistent with the others. Thanks for letting me know!

File: portfolio/src/main/java/com/google/sps/data/Constants.java [hunk 41]
```diff
@@ -1,5 +1,8 @@
 package com.google.sps.data;
 
+/*
```
Comment (razvanm) [inline portfolio/src/main/java/com/google/sps/data/Constants.java]:
```suggestion
/**
```

**Review by razvanm — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/data/Constants.java (hunk 41 above)
Comment (carolksun) [reply portfolio/src/main/java/com/google/sps/data/Constants.java]: fixed!

**Review by razvanm — APPROVED** [review]
